### PR TITLE
feature:support transaction propagation for globalTransaction api

### PR DESCRIPTION
**Pull request by funky-eyes**
### Ⅰ. Describe what this PR did
support transaction propagation for api

### Ⅱ. Does this pull request fix one issue?

#2375

### CODE REVIEW

**Comment by a364176773** [comment]
@zjinlei @slievrly @xingfudeshi @objcoding PTAL

File: tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java [hunk 1]
```diff
@@ -108,8 +113,34 @@ public void begin(int timeout, String name) throws TransactionException {
...
 
+    @Override
+    public void begin(int timeout, String name, Propagation propagation) throws TransactionException {
+        this.propagation = propagation;
+        switch (this.propagation) {
+            case NOT_SUPPORTED:
+                RootContext.unbind();
+                return;
+            case REQUIRES_NEW:
+                RootContext.unbind();
```
Comment (slievrly) [inline tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java]: globalTransaction1.begin(6000,"xxx1");
DML1
globalTransaction1.begin(6000,"xxx2", Propagation. REQUIRES_NEW);
DML2
globalTransaction2.commit;
globalTransaction1.commit;    where is the xid1?

File: tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java [hunk 2]
```diff
@@ -88,6 +97,30 @@ public void begin(int timeout) throws TransactionException {
 
     @Override
     public void begin(int timeout, String name) throws TransactionException {
+        begin(timeout, name, Propagation.REQUIRED);
+    }
+
+    @Override
+    public void begin(int timeout, String name, Propagation propagation) throws TransactionException {
+        isUsed();
+        this.propagation = propagation;
```
Comment (lightClouds917) [inline tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java]: Do we need to restore this.propagation?

Comment (a364176773) [reply tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java]: > Do we need to restore this.propagation?

no,the same instance is allowed to be used only once

File: tm/src/test/java/io/seata/tm/api/TransactionTemplateTest.java [hunk 3]
```diff
@@ -32,9 +32,7 @@
 import java.util.Set;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
Comment (slievrly) [inline tm/src/test/java/io/seata/tm/api/TransactionTemplateTest.java]: forbid to use *

File: tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java [hunk 4]
```diff
@@ -105,11 +138,13 @@ public void begin(int timeout, String name) throws TransactionException {
         if (LOGGER.isInfoEnabled()) {
             LOGGER.info("Begin new global transaction [{}]", xid);
         }
-
     }
 
     @Override
     public void commit() throws TransactionException {
+        if (Propagation.NOT_SUPPORTED.equals(this.propagation)) {
+            return;
```
Comment (slievrly) [inline tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java]: how to resume the pre xid?

File: tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java [hunk 5]
```diff
@@ -145,6 +181,9 @@ public void commit() throws TransactionException {
 
     @Override
     public void rollback() throws TransactionException {
+        if (Propagation.NOT_SUPPORTED.equals(this.propagation)) {
+            return;
```
Comment (slievrly) [inline tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java]: how to resume the pre xid?

Comment (a364176773) [reply tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java]: > how to resume the pre xid?

PTAL

File: tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java [hunk 6]
```diff
@@ -154,16 +202,19 @@ public void rollback() throws TransactionException {
             }
             return;
         }
-        assertXIDNotNull();
-
-        int retry = ROLLBACK_RETRY_COUNT;
         try {
+            if (Propagation.NOT_SUPPORTED.equals(this.propagation)) {
+                return;
```
[the remark this answers is not in the document]
Comment (a364176773) [reply tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java]: PTAL @lightClouds917

**Comment by a364176773** [comment]
please try again for ci

**Comment by a364176773** [comment]
please try again for ci @slievrly

File: tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java [hunk 7]
```diff
@@ -90,6 +97,44 @@ public void begin(int timeout) throws TransactionException {
...
+    }
+
+    @Override
+    public void begin(int timeout, String name, Propagation propagation) throws TransactionException {
+        assertInUsed();
+        this.propagation = propagation;
+        switch (propagation) {
+            case NOT_SUPPORTED:
+                suspendedResourcesHolder = suspend(true);
+                return;
```
[the remark this answers is not in the document]
Comment (a364176773) [reply tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java:113]: Shouldn't that be the case?

File: tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java [hunk 8]
```diff
@@ -90,6 +97,44 @@ public void begin(int timeout) throws TransactionException {
...
+            case REQUIRES_NEW:
+                suspendedResourcesHolder = suspend(true);
+                break;
+            case SUPPORTS:
+                if (!existingTransaction()) {
+                    return;
```
[the remark this answers is not in the document]
Comment (a364176773) [reply tm/src/main/java/io/seata/tm/api/DefaultGlobalTransaction.java:119]: > you just return this begin() method,then,The TransactionTemplate.execute method will continue to execute.

Shouldn't that be the case?

**Pull request closed without merging**